### PR TITLE
Warning on always true/false expressions involving NaN

### LINKED ISSUE #15936: != double.NaN should be a warning, and suggestion to fix
**Version Used**: 
Roslyn Language Services
2.0.0.6112806

**Steps to Reproduce**:
Write code:
double a = Foo();
if (a != double.NaN)
{
Thing();
[Program.txt](https://github.com/dotnet/roslyn/files/655721/Program.txt)
}
else
{
OtherThing();
}

**Expected Behavior**:
Should see a warning on != double.NaN saying "This expression is always true, consider !IsNaN(a)", perhaps a code refactoring to actually make that change, and a warning on OtherThing(); complaining about unreachable code.

**Actual Behavior**:
None of the above occur, and the code builds without warnings.
[Program.zip](https://github.com/dotnet/roslyn/files/655724/Program.zip)

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #15936

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:560]: I'll extract this logic and also extend it with float.NaN.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:562]: I'll consider not using magic strings here.

**Comment by Youssef1313** [comment]
I'm not sure where the tests for this should go 😄

**Comment by 333fred** [comment]
While I do think this would be a good warning for a warning wave, we're trying to make sure we don't add so many that people consider it a blocker because they get hit with a warning wall on upgrading. We already have 6 (counting the definite assignment warnings under one warning), which is about the limit of what we had previously discussed. @jaredpar, what do you think? Hold off on this until the next wave?

**Comment by CyrusNajmabadi** [comment]
If gate it by how many hours we expect people to have. Honestly, this warning seems fine to me. Most people will never see it. And the few that do will likely benefit. 

Note: for something like this, I think we should ship with a fixer as well. The more fixers we have for warning waves, the more palatable they will be. 

Fred, can you link me to your list of warnings you are shipping?  Thanks!

**Comment by 333fred** [comment]
@CyrusNajmabadi https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/Warnversion%20Warning%20Waves.md

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:562]: It seems like `node.OperatorToken` would be good here?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:562]: @RikkiGibson yes I think it should be the good approach here.
I'll get to that if it was decided to include this warning in the current warning wave (level 5).

It also seems that the test passes in CI, while it should fail because I haven't passed the expected diagnostics to `VerifyDiagnostics` yet. Not sure how to debug this issue if the call to `Error(diagnostics, ErrorCode.WRN_EqualityWithNaN, ....);` is hit.

**Comment by huoyaoyuan** [comment]
`float.NaN` should be covered together with double.
How about `Half.NaN`? I think the best solution is wait until we have concepts/type classes, and do the warning on `concept IEEE754FloatingPointNumber`.

**Comment by CyrusNajmabadi** [comment]
I don't think there's any need to wait on this. We can help people now with very low cost, and waiting longer won't end up really helping anyone.

**Comment by Youssef1313** [comment]
I just discovered there is CA2242 for this purpose.
Do we still need this warning on that case?

**Comment by RikkiGibson** [comment]
> 
> 
> I just discovered there is CA2242 for this purpose.
> Do we still need this warning on that case?

I'm not yet sure what we want to do in a scenario like this. @jmarolf @mavasani is there any interest in effectively moving a diagnostic+fix from roslyn-analyzers to roslyn warning waves and the built-in IDE fixers?

**Comment by mavasani** [comment]
CA2242 should be enabled by default as an IDE suggestion starting.NET5 SDK, I don’t think it needs a duplicate compiler warning. In fact @jmarolf even brought up turning CA2242 into a warning by default, can’t recall our resolution on that one.

**Comment by Youssef1313** [comment]
If it's enabled by default as a warning, I absolutely think it's redundant to add it as a compiler warning, especially that most likely, the compiler won't catch any cases that the analyzers can't.

Customers may even complain about duplicated warnings.

**Comment by mavasani** [comment]
I think we decided to leave it as an IDE suggestion by default to be conservative. Let’s avoid adding a duplicate diagnostic and instead move the discussion to the correct default severity for CA2242.

**Comment by jmarolf** [comment]
CA2242 was approved by Stephen Toub and friends as being on by default. The only issue we had (which Neil raised) is that the codefix does not use pattern matching (which would be the modern way to fix this issue).

given:
```csharp
using System;
public class C {
    public bool  M(double d) {
        return d == double.NaN;
    }
}
```

Code fix offers to change it to
```csharp
using System;
public class C {
    public bool  M(double d) {
        return double.IsNaN(d);
    }
}
```

If the language version is higher than 7 we would prefer this
```csharp
using System;
public class C {
    public bool  M(double d) {
        return d is double.NaN;
    }
}
```

Still, I am _very_ ok with this going into the compiler and being on by default in .NET 5. @333fred if the compiler team has no objections, this is fine to go in. 

@mavasani is there any reason we want this to live in the roslyn-analyzers repo vs. just being in the compiler? Considering the compiler is aware of `d is double.NaN` and translates it into `double.IsNaN(d)` this feels like a language level issue and not a library issue. Today doing this comparison is always incorrect and I would be very surprised if there was pushback from users here.

**Comment by mavasani** [comment]
> @mavasani is there any reason we want this to live in the roslyn-analyzers repo vs. just being in the compiler? Considering the compiler is aware of d is double.NaN and translates it into double.IsNaN(d) this feels like a language level issue and not a library issue. Today doing this comparison is always incorrect and I would be very surprised if there was pushback from users here.

It is a wrong statement to group CA analyzers as just .NET API analyzers. `Microsoft.CodeAnalysis.NetAnalyzers` inserted into the SDK has two big analysis areas:

1. .NET API analyzers
2. Generic code-quality analyzers, which are API agnostic but generic code quality improvement analyzers.

`CA2242` falls into the second bucket. It has existed for close to a decade, starting in FxCop, then ported to FxCop analyzers and now .NET SDK analyzers. I don't see a good reason to deprecate this ID and move it to the compiler - it seems like lot of unnecessary work for no net gain. I am going to mark this PR as blocked until both the IDE team and compiler team are able to approve that we want to go the route of moving API agnostic CA analyzers into compiler layer.

**Review by mavasani — CHANGES_REQUESTED** [review]
As per last comment.

**Comment by mavasani** [comment]
> CA2242 was approved by Stephen Toub and friends as being on by default. The only issue we had (which Neil raised) is that the codefix does not use pattern matching (which would be the modern way to fix this issue).

If that was the only blocking issue, it should be pretty easy to write/adjust the code fix for CA2242. It would be best to focus energy on that, as there is a customer value add with that work.

**Comment by jmarolf** [comment]
@mavasani I am also ok with us just shipping `CA2242` as on by default. I do not think this is a generic code-quality analyzer. Writing `d == double.NaN` is always incorrect. There is no situation where the user should not change their code. If we feel the history of `CA2242` means we want to keep this around I am fine with that but I generally want warnings to come from the compiler if possible for performance reasons.

**Comment by mavasani** [comment]
That’s a reasonable counter argument Jon. If we were to do this from scratch then certainly a compiler warning would be more sensible. I’ll let the compiler team decide on how to proceed on this one. My block is mainly to ensure we have their design approval before moving forward on this PR.

**Comment by jmarolf** [comment]
Totally agree, if the compiler doesn't want this then I think we should enable the analyzer

**Comment by jaredpar** [comment]
> We already have 6 (counting the definite assignment warnings under one warning), which is about the limit of what we had previously discussed. @jaredpar, what do you think? Hold off on this until the next wave?

Agree with @333fred here. Additionally we need to get LDM approval that this is a warning we want to include in a future wave. At the moment we are still essentially using LDM to approve warning wave additions since it's a core part of the language.

**Comment by Youssef1313** [comment]
@jaredpar Should I close? or this will be discussed in a future LDM?

**Comment by 333fred** [comment]
> jaredpar Should I close? or this will be discussed in a future LDM?

Let's leave it open and bring it to a future LDM. I'll add it to the schedule.

**Comment by jcouv** [comment]
Marked PR as draft until we hear back from LDM.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (jkotas) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:559]: Can this apply to `<`, `>`, `<=` and `>=` as well? (See "Comparing with NaN" in https://en.wikipedia.org/wiki/NaN).

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:559]: @jkotas Sure. I'll update it if the warning is approved in future LDM.

**Comment by 333fred** [comment]
LDM met on this today, and the decision was keep this as an FxCop warning. People have already had to deal with this warning there, and have suppressed it if they really want to do the comparison anyway. Since the analyzers are shipping with .NET 5 itself, we don't see any real need to introduce warning number churn for the sole purpose of bringing this into the compiler itself.

**Pull request closed without merging**
